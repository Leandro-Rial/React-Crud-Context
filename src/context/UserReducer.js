import { types } from "../types/types"

export const UserReducer = (state = {}, action) => {
    switch (action.type) {
        case types.removeUser:
            
            return {
                ...state,
                users: state.users.filter(user => {
                    return user.id !== action.payload
                })
            }
        
        case types.addUser:
            
            return {
                ...state,
                users: [action.payload, ...state.users]
            }
        
        
        case types.editUser:
            const updateUser = action.payload;
            
            const updateUsers = state.users.map(user => {
                if(user.id === updateUser.id) {
                    return updateUser
                }

                return user;
            })

            return {
                ...state,
                users: updateUsers
            }
        
    
        default:
            return state
    }
}