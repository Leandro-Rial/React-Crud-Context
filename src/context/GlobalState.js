import React, { createContext, useEffect, useReducer } from 'react';
import { types } from '../types/types';
import { UserReducer } from './UserReducer';

const initialState = {
    users: []
}

const init = () => {
    return (
        JSON.parse(localStorage.getItem("users")) || initialState
    )
}

export const UserContext = createContext(initialState);

export const UserProvider = ({ children }) => {

    const [state, dispatch] = useReducer(UserReducer, initialState, init)

    useEffect(() => {
        localStorage.setItem('users', JSON.stringify(state))
    }, [state])

    const removeUser = (id) => {
        dispatch({
            type: types.removeUser,
            payload: id
        })
    }
    
    const addUser = (user) => {
        dispatch({
            type: types.addUser,
            payload: user,
        })
    }

    const editUser = (user) => {
        dispatch({
            type: types.editUser,
            payload: user
        })
    }

    return (
        <UserContext.Provider value={{
            users: state.users,
            removeUser,
            addUser,
            editUser
        }}>
            {children}
        </UserContext.Provider>
    )
    
}