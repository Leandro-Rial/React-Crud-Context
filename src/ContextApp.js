import React from 'react';
import { UserProvider } from './context/GlobalState';
import AppRouter from './router/AppRouter';

function App() {
  return (
    <UserProvider>
      <AppRouter />
    </UserProvider>
  );
}

export default App;
