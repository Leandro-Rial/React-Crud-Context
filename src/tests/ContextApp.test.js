import { shallow } from "enzyme";
import { UserReducer } from "../context/UserReducer";
import ContextApp from '../ContextApp'

describe('Prubas en ContextApp', () => {
    test('debe de mostarse correctamente', () => {
        const wrapper = shallow(
            <UserReducer>
                <ContextApp />
            </UserReducer>
        )

        expect(wrapper).toMatchSnapshot()
    });
});