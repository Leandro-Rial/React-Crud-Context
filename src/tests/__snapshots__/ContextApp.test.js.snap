// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prubas en ContextApp debe de mostarse correctamente 1`] = `<Component />`;
